File: backend/scripts/convert_csv.py
import pandas as pd
import json
import os
import math

source_file = os.path.join(os.path.dirname(__file__), '../../data-pipeline/raw/Recipes_with_Images.csv')
target_file = os.path.join(os.path.dirname(__file__), '../data/recipes_small.json')

print(f"Reading from {source_file}")
df = pd.read_csv(source_file, index_col=False)

recipes = []
for index, row in df.iterrows():
    # Handle NaN values safely
    def get_val(val, default=''):
        return default if pd.isna(val) else val
        
    def split_lines(txt):
        if pd.isna(txt): return []
        return [line.strip() for line in str(txt).split('\n') if line.strip()]

    recipe = {
        "id": str(get_val(row.get('index'), index)),
        "title": get_val(row.get('title'), 'Unknown Recipe').strip(),
        "about": get_val(row.get('about')),
        "ingredients": split_lines(row.get('ingridients')),
        "directions": split_lines(row.get('preparation')),
        "imageUrl": None,
        "prepTime": 30,
        "cuisine": "African",
        "difficulty": "Medium",
        "nutrition": {
            "energy_kcal": get_val(row.get('energy(kcal)'), 0),
            "fat_g": get_val(row.get('fat(g)'), 0),
            "carbohydrates_g": get_val(row.get('carbohydrates(g)'), 0),
            "proteins_g": get_val(row.get('proteins(g)'), 0),
            "fibre_g": get_val(row.get('fibre(g)'), 0)
        }
    }
    
    # Process image path
    image_path = get_val(row.get('image_path'))
    if image_path:
        # Expected format: "images/..."
        recipe['imageUrl'] = f"/{image_path}"
        
    recipes.append(recipe)

output = {
    "recipes": recipes
}

os.makedirs(os.path.dirname(target_file), exist_ok=True)
with open(target_file, 'w', encoding='utf-8') as f:
    json.dump(output, f, indent=2)

print(f"Successfully wrote {len(recipes)} recipes to {target_file}")

File: backend/scripts/convert_csv.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import Papa from 'papaparse';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const SOURCE_FILE = path.join(__dirname, '../../data-pipeline/raw/Recipes_with_Images.csv');
const TARGET_FILE = path.join(__dirname, '../data/recipes_small.json');

console.log(`Starting conversion of ${SOURCE_FILE}...`);

try {
  const rawData = fs.readFileSync(SOURCE_FILE, 'utf-8');

  // Parse CSV
  const parsed = Papa.parse(rawData, {
    header: true,
    skipEmptyLines: true,
  });

  if (parsed.errors.length) {
    console.warn(`CSV Parsing warnings:`, parsed.errors);
  }

  const recipes = parsed.data.map((row) => {
    return {
      id: row.index ? row.index.toString() : Math.random().toString().slice(2, 8),
      title: row.title?.trim() || 'Unknown Recipe',
      about: row.about,
      // The CSV has "ingridients" instead of "ingredients"
      ingredients: row.ingridients
        ? row.ingridients
            .split('\n')
            .map((i) => i.trim())
            .filter(Boolean)
        : [],
      directions: row.preparation
        ? row.preparation
            .split('\n')
            .map((p) => p.trim())
            .filter(Boolean)
        : [],
      imageUrl: row.image_path ? `/api/${row.image_path}` : null,
      prepTime: 30, // Default prep time
      cuisine: 'Kenyan', // Based on the recipes like Kaimati, Mandazi
      difficulty: 'Medium',
      nutrition: {
        energy_kcal: row['energy(kcal)'],
        fat_g: row['fat(g)'],
        carbohydrates_g: row['carbohydrates(g)'],
        proteins_g: row['proteins(g)'],
        fibre_g: row['fibre(g)'],
      },
    };
  });

  const output = {
    recipes: recipes,
  };

  const dataDir = path.dirname(TARGET_FILE);
  if (!fs.existsSync(dataDir)) {
    fs.mkdirSync(dataDir, { recursive: true });
  }

  fs.writeFileSync(TARGET_FILE, JSON.stringify(output, null, 2));
  console.log(`Successfully converted ${recipes.length} recipes to ${TARGET_FILE}!`);
} catch (error) {
  console.error('Error extracting sample:', error.message);
}
